flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -1243,6 +1250,29 @@ private void completePendingCheckpoint(PendingCheckpoint pendingCheckpoint)
                         CheckpointFailureReason.FINALIZE_CHECKPOINT_FAILURE,
                         exception);
             }
+
+            try {
+                allPreviousCheckpoints.removeAll(completedCheckpointStore.getAllCheckpoints());
+                if (!allPreviousCheckpoints.isEmpty()) {
```

Comment:
The original idea is to not change the API, and I think a proper place is to change the return result of `CompletedCheckpointStore#addCheckpoint` from `void` to the subsumed `CompletedCheckpoint`.

Reply:
I think returning **subsumed** checkpoint from `addCheckpoint` is not very straightforward (as a caller, I'd expect **added** checkpoint to be returned).
I'd rather add `long getLatestSubsumedCheckpointID();` method or something like that.

Reply:
Actually, current `CheckpointStore#addCheckpoint` has the meaning that this is an atomic operation which will add the completed checkpoint and subsumed an older checkpoint.` I think return the subsumed checkpoint is clearly for this operation.
If we adopt the API `long getLatestSubsumedCheckpointID()`, we will miss some information that when the checkpoint is subsumed and we might also need to check whether the latest subsumed checkpoint ID has been changed.

Reply:
> Actually, current CheckpointStore#addCheckpoint has the meaning that this is an atomic operation which will add the completed checkpoint and subsumed an older checkpoint. I think return the subsumed checkpoint is clearly for this operation.

To me, it's only clear that it might subsume some older checkpoints. It is unclear whether returned is the first subsumed, the latest subsumed, or the added one.

> If we adopt the API long getLatestSubsumedCheckpointID()`, we will miss some information that when the checkpoint is subsumed and we might also need to check whether the latest subsumed checkpoint ID has been changed.

I think there is no requirement for the CheckpointCoordinator to report strictly latest ID (it will be outdated right after the RPC). It must only be non-decreasing. If it somehow becomes greater than the added one we could use the smallest from them.

And currently Store is used from the same thread, so calling `addCheckpoint` and then `getLatestSubsumed` is safe (I don't think this concurrency model should be changed).

Reply:
I think we can refer to the documentation of `CompletedCheckpointStore#addCheckpoint`:
~~~ java
Only a bounded number of checkpoints is kept. When exceeding the maximum number of
retained checkpoints, the oldest one will be discarded.
~~~
And all current implementations obey this role to make this method atomic and discard the oldest checkpoint internally. That's why, to me, changing this method to return the oldest discarded checkpoint looks so straightforward.

Reply:
Yes, it's clear when one goes through the documentation. 
However, I think it's better to make the API clear even without the need to read the javadoc.

Reply:
I see, maybe the previous name of `addCheckpoint` is not clear for what it really does. It actually does `addCheckpointAndSubsumeOldestOne`.

Reply:
The new name doesn't seem to reflect what the method actually does.
It doesn't necessarily subsume **one** checkpoint (or **any** checkpoints at all).

The correct name would be `addAndMaybeSubsumeOldestAndReturnTheLatestSubsumedIfAny` - which clearly indicates to me that the new functionality should be in a separate method.

Reply:
Yes, I agree that the returned subsumed checkpoint might be null, but I can't see why we need to add a separate method? If so, who will call the original `addCheckpoint` method?

Reply:
> Yes, I agree that the returned subsumed checkpoint might be null

From the (current) name, it's unclear which checkpoint is returned, even if not null.

> I can't see why we need to add a separate method? 
> who will call the original addCheckpoint method?

In my opinion, it would make a more clear API.
`CheckpointCoordinator` will call it (same as now). Let me demonstrate this: 
https://github.com/rkhachatryan/flink/commit/10242b7aef2f4f2db94c924f940928fce4ab0fc9#diff-2c7004a2d412c3566de5ff6fb9e6d027742328c2acad60685e47ce4ba9df0810R1286

WDYT?